### PR TITLE
Limit typing indicators

### PR DESCRIPTION
**Pull request by grzesiekko**
### What?

- Typing indicators displayed by `TypingUsersViewModel` limited to 3.
- Data simulation no longer sends typing indicators of the currently logged in user.

### Why?

User testing.

----

### CODE REVIEW

**Comment by grzesiekko** [comment]
Actually, I might add 1 thing here before merging.

**Comment by grzesiekko** [comment]
Done. The PR can be reviewed now.

**Comment by mdpye** [comment]
Hah, I think you've really hit on something here!

I believe that you do receive your own typing indicators from the web service, so rather than removing them from the dummy data, we should probably be filtering the current user from the list in the view model...

**Comment by grzesiekko** [comment]
I will add such functionality.

File: Chatkit/Data Providers/TypingUsersProvider.swift [hunk 1]
```diff
@@ -27,10 +27,13 @@ public class TypingUsersProvider {
...
     }
     
+    let currentUser: User
+    
     // MARK: - Initializers
     
-    init(room: Room, persistenceController: PersistenceController) {
+    init(room: Room, currentUser: User, persistenceController: PersistenceController) {
         self.roomIdentifier = room.identifier
+        self.currentUser = currentUser
```
Comment (mdpye) [inline Chatkit/Data Providers/TypingUsersProvider.swift]: As the other fields of user (name, custom data etc) might change, and won't be updated here, perhaps we should only store the id, as it's all we actually need to identify the user?

Comment (mdpye) [reply Chatkit/Data Providers/TypingUsersProvider.swift]: Also, surely this should be registered in the ViewModel where it is used, not here?

If you're worried about the constructor args for the view model, I think it's either valid to require it, or we can add factory methods for view models to the Chatkit object as Zan suggested

Comment (grzesiekko) [reply Chatkit/Data Providers/TypingUsersProvider.swift]: One could argue that the same is valid for `typingUsers` property of `TypingUsersProvider`.

Comment (mdpye) [reply Chatkit/Data Providers/TypingUsersProvider.swift]: Aren't the typing users snapshot from the coredata frequently, at least? (but yes, that is a conversation we still need to have!)

File: Chatkit/Data Providers/TypingUsersProvider.swift [hunk 2]
```diff
@@ -27,10 +27,13 @@ public class TypingUsersProvider {
         return Set(users)
     }
     
+    let currentUserIdentifier: String
```
Comment (mdpye) [inline Chatkit/Data Providers/TypingUsersProvider.swift]: I still don't see that this at belonging in here when it's not used here

Comment (grzesiekko) [reply Chatkit/Data Providers/TypingUsersProvider.swift]: OK, in that case should we instantiate the view model with an instance of `Chatkit` in order to be sure that the user passed to the view model is actually the one that is currently logged into the system?

Comment (grzesiekko) [reply Chatkit/Data Providers/TypingUsersProvider.swift]: It look like we have 3 options here:

1. Leave it as it is right now and hide the `currentUser` from developers:
   - The public interface is simpler.
   - It is not possible to instantiate the view model with an incorrect instance of `User`.
   - Unfortunately, the identifier needs to be passed here somehow (internally).
2. Instantiate the view model with an instance of `Chatkit`:
   - The public interface is a bit more complex simpler.
   - It is still not possible to instantiate the view model with an incorrect instance of `User` (that's a good thing).
   - It **is** possible to instantiate the view model with an incorrect instance of `Chatkit`.
   - Initializer of the view model will have to be be optional (error handling would be optional as well).
3. Instantiate the view model with an instance of `User` or an identifier of the current user (`String`):
   - The public interface is a bit more complex simpler.
   - It **is** possible to instantiate the view model with an incorrect instance of `User` (that's not a good thing).
   - A bit cleaner implementation (internally).

Comment (mdpye) [reply Chatkit/Data Providers/TypingUsersProvider.swift]: Hrm. I see what you mean. Personally, I'd definitely request it from the TypingIndicatorsViewModel constructor - I was even thinking to turn it in to a generic "filter user" parameter. I'm unsure about the correctness / ease of use tradeoff here.

There is a 4th option, which Zan suggested:
- create view models *from* the Chatkit object, like we do with providers.

I think this is definitely the most ergonomic, but it also makes it seem more "magic" to the customer, which would discourage them (or make a different way of implementing at least) their own view models in future.

Comment (grzesiekko) [reply Chatkit/Data Providers/TypingUsersProvider.swift]: After the recent change all view models are now being initialized using factory methods.

**Pull request merged**